### ISSUE TITLE
Clean up quotient

### ISSUE THREAD

**Issue by Alizter** [HoTT/HoTT#1088]
I would like to propose cleaning up quotient.v. This would involve renaming the file, renaming some of the lemmas to adhere to conventions. It may also be worth considering if we want to define quotients as a coequalizer? We can keep comments detailing the private inductive version too.

What are your thoughts? Do you have anything to add?

**Comment by spitters** [comment]
I don't recall progress on the private inductive types, but @andreaslyn has been running into some problems when porting math-components to HoTT. @ppedrot is experimenting with this, so he may have some insights.

**Comment by Alizter** [comment]
I am a bit confused why this is happening. It seems to me, correctly annotating quotient actually causes `V.v` to fail, which is why Mike couldn't annotate the private inductive type back then. Now that I have defined it as a truncated coequalizer, we have the annotations too, so this seems to be the same problem.

I don't think this actually has anything to do with private inductive types, but rather `V.v` might have some sort of mismatch with universes that we weren't seeing when using the private inductive type since it just let things happen.

**Comment by mikeshulman** [comment]
Someone should look at it carefully.  Unfortunately I don't have time now...

**Comment by Alizter** [comment]
I found a mistake with my definition: I used `Tr` instead of `Trunc` which meant that `quotient` was living inside `Typele` rather than `Type`. This hasn't fixed `V.v` however...

In ln 438 of V.v:
```coq
(** This lemma actually says a little more than 10.5.6, i.e., that Au is a hSet *)
Lemma monic_set_present : forall u : V, exists (Au : Type) (m : Au -> V),
  (IsHSet Au) * (IsEmbedding m) * (u = set m).
Proof.
  apply V_ind_hprop.
  - intros A f _.
    destruct (quotient_kernel_factor f (ker_bisim f) (ker_bisim_is_ker f))
      as [Au [eu [mu (((hset_Au, epi_eu), mono_mu), factor)]]].
    exists Au, mu. split;[exact (hset_Au, mono_mu)|].
    apply setext'; split.
    + intro a. apply tr; exists (eu a). exact (ap10 factor a).
    + intro a'. generalize (epi_eu a').
      intros ?; refine (Trunc_functor -1 _ (center _)).
      intros [a p]. exists a. transitivity (mu (eu a)).
      * exact (ap10 factor a).
      * exact (ap mu p).
  - intro v. apply hprop_allpath.
    intros [Au [mu ((hset, mono), p)]].
    intros [Au' [mu' ((hset', mono'), p')]].
    apply monic_set_present_uniqueness.
Defined.
```

We can't get passed `exists Au` since:

```coq
In environment
ua : Univalence
A : Type
f : A -> V
Au : Type
eu : A -> Au
mu : Au -> V
hset_Au : IsHSet Au
epi_eu : IsSurjection eu
mono_mu : IsEmbedding mu
factor : f = (fun x : A => mu (eu x))
The term "Au" has type "Type@{HoTT.HIT.V.11892}" while it is expected to have type
 "Type@{HoTT.HIT.V.10895}" (universe inconsistency).
```

It must be a problem with `quotient_kernel_pair` in quotient.v (since I changed it):
```coq


Section Kernel.

  (** ** Quotients of kernels of maps to sets give a surjection/mono factorization. *)

  Context `{Funext}.

  (** A function we want to factor. *)
  Context {A B : Type} `{IsHSet B} (f : A -> B).

  (** A mere relation equivalent to its kernel. *)
  Context (R : relation A) {sR : is_mere_relation _ R}
          (is_ker : forall x y, f x = f y <~> R x y).

  Theorem quotient_kernel_factor
    : exists (C : Type) (e : A -> C) (m : C -> B),
      IsHSet C * IsSurjection e * IsEmbedding m * (f = m o e).
  Proof.
    pose (C := quotient R).
    (* We put this explicitly in the context so that typeclass resolution will pick it up. *)
    assert (IsHSet C) by (unfold C; apply _).
    exists C.
    pose (e := class_of R).
    exists e.
    transparent assert (m : (C -> B)).
    { apply quotient_ind with f; try exact _.
      intros x y p.
      transitivity (f x).
      - apply transport_const.
      - exact ((is_ker x y)^-1 p). }
    exists m.
    do 3 try split; try exact _.
    intro u.
    apply hprop_allpath.
    assert (p : forall (x y : C) (p : m x = u) (p' : m y = u), x = y).
    { simple refine (quotient_ind R _ _ _).
      - intro a.
        simple refine (quotient_ind R _ _ _).
        + intros a' p p'; fold e in p, p'.
          * apply related_classes_eq.
            refine (is_ker a a' _).
            change (m (e a) = m (e a')).
            exact (p @ p'^).
        + intros; apply path_ishprop.
      - intros; apply path_ishprop. }
    intros [x q] [y p'].
    apply path_sigma_hprop; simpl.
    exact (p x y q p').
  Defined.

End Kernel.
```

Here is the full file if you want to try:
```coq
Require Import Basics.
Require Import Types.
Require Import HSet.
Require Import HProp.
Require Import UnivalenceImpliesFunext.
Require Import TruncType.
Require Import HIT.epi HIT.Truncations HIT.Connectedness.

Local Open Scope path_scope.

(* TODO: Move *)
Definition apD_compose {A A' : Type} {B : A' -> Type} (g : A -> A')
  (f : forall a, B a) {x y : A} (p : x = y)
  : apD (f o g) p  = (transport_compose _ _ _ _) @ apD f (ap g p).
Proof.
  by destruct p.
Defined.

(* TODO: Move *)
Definition apD_compose' {A A' : Type} {B : A' -> Type} (g : A -> A')
  (f : forall a, B a) {x y : A} (p : x = y)
  : apD f (ap g p) = (transport_compose B _ _ _)^ @ apD (f o g) p.
Proof.
  by destruct p.
Defined.

(** * Quotient of a Type by an hprop-valued relation

We aim to model:
<<
Inductive quotient : Type :=
   | class_of : A -> quotient
   | related_classes_eq : forall x y, (R x y), class_of x = class_of y
   | quotient_set : IsHSet quotient
>>
*)

(** This development should be further connected with the sections in the book; see below.*)

(* TODO: Move up *)
Require Import HIT.Coeq.

Section Quotient.

  (* TODO: Rename, clean up *)
  Definition quotient {A : Type} (R : relation A)
    `{is_mere_relation _ R}
    := Trunc 0 (Coeq 
        (fun x : {a : A & {b : A & R a b}} => x.1)
        (fun x : {a : A & {b : A & R a b}} => x.2.1)).

  (* TODO: rename *)
  Definition class_of {A : Type} (R : relation A)
    `{is_mere_relation _ R} : A -> quotient R := tr o coeq.

  (* TODO: rename *)
  Definition related_classes_eq {A : Type} {R : relation A}
    `{is_mere_relation _ R} {x y : A}
    : R x y -> class_of R x = (class_of R y : quotient R)
    := fun p => ap tr (cp (x; y; p)).

  Context {A : Type} (R : relation A) `{is_mere_relation _ R}.

  (* TODO: rename *)
  Definition quotient_ind (P : quotient R -> Type)
    {sP : forall x, IsHSet (P x)} (pclass : forall x, P (class_of R x))
    (peq : forall x y (H : R x y), related_classes_eq H # pclass x = pclass y)
    : forall q, P q.
  Proof.
    eapply Trunc_ind, Coeq_ind.
    1: assumption.
    intros [a [b p]].
    refine (transport_compose _ _ _ _ @ _).
    apply peq.
  Defined.

  (* TODO: rename *)
  Definition quotient_rec (P : Type) `{IsHSet P} (pclass : A -> P)
    (peq : forall x y, R x y -> pclass x = pclass y)
    : quotient R -> P.
  Proof.
    apply Trunc_rec.
    serapply Coeq_rec.
    1: assumption.
    intros [a [b p]].
    apply peq.
    assumption.
  Defined.

  (* TODO: rename *)
  Definition quotient_ind_compute_path (P : quotient R -> Type)
    {sP : forall x, IsHSet (P x)} (pclass : forall x, P (class_of R x))
    (peq : forall x y (H : R x y), related_classes_eq H # pclass x = pclass y)
    : forall x y p, apD (quotient_ind P pclass peq) (related_classes_eq p)
      = peq x y p.
  Proof.
    intros x y p.
    unfold related_classes_eq.
    refine (apD_compose' tr _ _ @ _).
    unfold quotient_ind.
    change ((transport_compose P tr (cp (x; y; p)) (pclass x))^ @
      apD (Coeq_ind (fun w => P (tr w)) pclass (fun b0 =>
      transport_compose P tr (cp (b0.1; (b0.2).1; (b0.2).2)) (pclass b0.1) @
      peq b0.1 (b0.2).1 (b0.2).2)) (cp (x; y; p)) = peq x y p).
    rewrite Coeq_ind_beta_cp.
    apply concat_V_pp.
  Defined.

  (* TODO: rename *)
  Definition quotient_rec_beta_related_classes_eq (P : Type) `{IsHSet P}
    (pclass : A -> P) (peq : forall x y, R x y -> pclass x = pclass y)
    : forall x y p, ap (quotient_rec P pclass peq) (related_classes_eq p)
      = peq x y p.
  Proof.
    intros x y p.
    rewrite <- (ap_compose tr).
    unfold quotient_rec.
    change (ap (Coeq_rec P pclass (fun b0 : {a : A & {b : A & R a b}} =>
      peq b0.1 (b0.2).1 (b0.2).2)) (cp (x; y; p)) = peq x y p).
    serapply Coeq_rec_beta_cp.
  Defined.

End Quotient.

Arguments quotient : simpl never.
Arguments class_of : simpl never.
Arguments related_classes_eq : simpl never.

Notation "A / R" := (quotient (A:=A) R).

(*
Module Export Quotient.

  Section Domain.

    Context {A : Type} (R:relation A) {sR: is_mere_relation _ R}.

    (** We choose to let the definition of quotient depend on the proof that [R] is a set-relations.  Alternatively, we could have defined it for all relations and only develop the theory for set-relations.  The former seems more natural.

We do not require [R] to be an equivalence relation, but implicitly consider its transitive-reflexive closure. *)


    (** Note: If we wanted to be really accurate, we'd need to put [@quotient A R sr] in the max [U_{sup(i, j)}] of the universes of [A : U_i] and [R : A -> A -> U_j].  But this requires some hacky code, at the moment, and the only thing we gain is avoiding making use of an implicit hpropositional resizing "axiom". *)

    (** This definition has a parameter [sR] that shadows the ambient one in the Context in order to ensure that it actually ends up depending on everything in the Context when the section is closed, since its definition doesn't actually refer to any of them.  *)
    Private Inductive quotient {sR: is_mere_relation _ R} : Type :=
    | class_of : A -> quotient.

    (** The path constructors. *)
    Axiom related_classes_eq
    : forall {x y : A}, R x y ->
                        class_of x = class_of y.

    Axiom quotient_set : IsHSet (@quotient sR).
    Global Existing Instance quotient_set.

    Definition quotient_ind (P : (@quotient sR) -> Type) {sP : forall x, IsHSet (P x)}
               (dclass : forall x, P (class_of x))
               (dequiv : (forall x y (H : R x y), (related_classes_eq H) # (dclass x) = dclass y))
    : forall q, P q
      := fun q => match q with class_of a => fun _ _ => dclass a end sP dequiv.

    Definition quotient_ind_compute {P sP} dclass dequiv x
    : @quotient_ind P sP dclass dequiv (class_of x) = dclass x.
    Proof.
      reflexivity.
    Defined.

    (** Again equality of paths needs to be postulated *)
    Axiom quotient_ind_compute_path
    : forall P sP dclass dequiv,
      forall x y (H : R x y),
        apD (@quotient_ind P sP dclass dequiv) (related_classes_eq H)
        = dequiv x y H.

  End Domain.

End Quotient. *)

Section Equiv.

  Context `{Univalence}.

  Context {A : Type} (R : relation A)
    `{is_mere_relation _ R}
    `{Transitive _ R}
    `{Symmetric _ R}.

  Lemma quotient_path2 : forall {x y : A / R} (p q : x=y), p=q.
  Proof.
    intros x y.
    apply set_path2.
  Defined.

  Definition in_class : A / R -> A -> hProp.
  Proof.
    serapply quotient_ind.
    1: intros a b; serapply (BuildhProp (R a b)).
    intros x y p.
    refine (transport_const _ _ @ _).
    funext z.
    apply path_hprop.
    serapply equiv_iff_hprop; cbn.
    1: apply (transitivity (symmetry _ _ p)).
    apply (transitivity p).
  Defined.

  Context `{Reflexive _ R}.

  Lemma in_class_pr : forall x y, (in_class (class_of R x) y : Type) = R x y.
  Proof.
    intro x.
    reflexivity.
  Defined.

  Lemma quotient_ind_hprop (P : quotient R -> Type) `{forall x, IsHProp (P x)}
    (dclass : forall x, P (class_of R x)) : forall q, P q.
  Proof.
    serapply (quotient_ind R P dclass).
    1: intro; apply trunc_succ.
    intros x y p.
    apply path_ishprop.
  Defined.

  Global Instance decidable_in_class `{forall x y, Decidable (R x y)}
  : forall x a, Decidable (in_class x a).
  Proof.
    serapply quotient_ind_hprop.
    intros a b; exact (transport Decidable (in_class_pr a b) _).
  Defined.

  Lemma class_of_repr : forall q x, in_class q x -> q = class_of R x.
  Proof.
    serapply quotient_ind.
    { intros x y p.
      apply (related_classes_eq p). }
    intros x y p.
    funext ? ?.
    apply quotient_path2.
  Defined.

  Lemma classes_eq_related : forall x y, class_of R x = class_of R y -> R x y.
  Proof.
    intros x y p.
    refine (transport idmap (in_class_pr  _ _) _).
    refine (transport (fun k => in_class k y) p^ _).
    simpl; reflexivity.
  Defined.

  (** Thm 10.1.8 *)
  Theorem sets_exact : forall x y, (class_of R x = class_of R y) <~> R x y.
  Proof.
    intros ??.
    apply equiv_iff_hprop.
    - apply classes_eq_related.
    - apply related_classes_eq.
  Defined.

  Definition quotient_rec2 {B : hSet} {dclass : A -> A -> B}
    {dequiv : forall x x', R x x' -> forall y y', R y y'
      -> dclass x y = dclass x' y'}
    : A / R -> A / R -> B.
  Proof.
    serapply quotient_rec.
    { intro a.
      serapply quotient_rec.
      { revert a.
        assumption. }
      by apply (dequiv a a). }
    intros x y p.
    apply path_forall.
    serapply quotient_ind.
    { simpl.
      intro a.
      by apply dequiv. }
    intros a b q.
    apply path_ishprop.
  Defined.

  Definition quotient_ind_hprop' (P : A / R -> Type)
    `{forall x, IsHProp (P (class_of _ x))}
    (dclass : forall x, P (class_of _ x)) : forall y, P y.
  Proof.
    apply quotient_ind with dclass.
    { serapply quotient_ind.
      1: intro; apply trunc_succ.
      intros ???; apply path_ishprop. }
    intros; apply path_ishprop.
  Defined.

  (** From Ch6 *)
  Global Instance quotient_surjective: IsSurjection (class_of R).
  Proof.
    apply BuildIsSurjection.
    serapply quotient_ind_hprop.
    intro x.
    apply tr.
    by exists x.
  Defined.

  (* Universal property of quotient *)
  Theorem equiv_quotient_ump (B : hSet)
    : (quotient R -> B) <~> {f : A -> B & forall x y, R x y -> f x = f y}.
  Proof.
    serapply equiv_adjointify.
    + intro f.
      exists (compose f (class_of R)).
      intros; f_ap.
      by apply related_classes_eq.
    + intros [f H'].
      apply (quotient_rec _ _ _ H').
    + intros [f Hf].
      by apply equiv_path_sigma_hprop.
    + intros f.
      apply path_forall.
      serapply quotient_ind_hprop.
      reflexivity.
  Defined.

  (** Missing

  The equivalence with VVquotient [A//R].

  This should lead to the unnamed theorem:

  10.1.10. Equivalence relations are effective and there is an equivalence [A/R<~>A//R]. *)

  (**
  The theory of canonical quotients is developed by C.Cohen:
  http://perso.crans.org/cohen/work/quotients/
 *)

End Equiv.

Section Functoriality.

  Definition quotient_functor
    {A : Type} (R : relation A) {sR: is_mere_relation _ R}
    {B : Type} (S : relation B) {sS: is_mere_relation _ S}
    (f : A -> B) (fresp : forall x y, R x y -> S (f x) (f y))
    : quotient R -> quotient S.
  Proof.
    refine (quotient_rec R _ (class_of S o f) _).
    intros x y r.
    apply related_classes_eq, fresp, r.
  Defined.

  Context {A : Type} (R : relation A) {sR: is_mere_relation _ R}
          {B : Type} (S : relation B) {sS: is_mere_relation _ S}.

  Global Instance quotient_functor_isequiv (f : A -> B)
    (fresp : forall x y, R x y <-> S (f x) (f y)) `{IsEquiv _ _ f}
    : IsEquiv (quotient_functor R S f (fun x y => fst (fresp x y))).
  Proof.
    serapply (isequiv_adjointify _ (quotient_functor S R f^-1 _)).
    { intros u v s.
      apply (snd (fresp _ _)).
      abstract (do 2 rewrite eisretr; apply s). }
    all: serapply quotient_ind.
    + intros b; simpl.
      apply ap, eisretr.
    + intros; apply path_ishprop.
    + intros a; simpl.
      apply ap, eissect.
    + intros; apply path_ishprop.
  Defined.

  Definition equiv_quotient_functor (f : A -> B) `{IsEquiv _ _ f}
    (fresp : forall x y, R x y <-> S (f x) (f y)) : quotient R <~> quotient S
    := BuildEquiv _ _ (quotient_functor R S f (fun x y => fst (fresp x y))) _.

  Definition quotient_functor_equiv' (f : A <~> B)
    (fresp : forall x y, R x y <-> S (f x) (f y))
    : quotient R <~> quotient S := equiv_quotient_functor f fresp.

End Functoriality.

Section Kernel.

  (** ** Quotients of kernels of maps to sets give a surjection/mono factorization. *)

  Context `{Funext}.

  (** A function we want to factor. *)
  Context {A B : Type} `{IsHSet B} (f : A -> B).

  (** A mere relation equivalent to its kernel. *)
  Context (R : relation A) {sR : is_mere_relation _ R}
          (is_ker : forall x y, f x = f y <~> R x y).

  Theorem quotient_kernel_factor
    : exists (C : Type) (e : A -> C) (m : C -> B),
      IsHSet C * IsSurjection e * IsEmbedding m * (f = m o e).
  Proof.
    pose (C := quotient R).
    (* We put this explicitly in the context so that typeclass resolution will pick it up. *)
    assert (IsHSet C) by (unfold C; apply _).
    exists C.
    pose (e := class_of R).
    exists e.
    transparent assert (m : (C -> B)).
    { apply quotient_ind with f; try exact _.
      intros x y p.
      transitivity (f x).
      - apply transport_const.
      - exact ((is_ker x y)^-1 p). }
    exists m.
    do 3 try split; try exact _.
    intro u.
    apply hprop_allpath.
    assert (p : forall (x y : C) (p : m x = u) (p' : m y = u), x = y).
    { simple refine (quotient_ind R _ _ _).
      - intro a.
        simple refine (quotient_ind R _ _ _).
        + intros a' p p'; fold e in p, p'.
          * apply related_classes_eq.
            refine (is_ker a a' _).
            change (m (e a) = m (e a')).
            exact (p @ p'^).
        + intros; apply path_ishprop.
      - intros; apply path_ishprop. }
    intros [x q] [y p'].
    apply path_sigma_hprop; simpl.
    exact (p x y q p').
  Defined.

End Kernel.
```

**Comment by Alizter** [comment]
I had a look and it seems this also closes #565. There was an effort to clean up universe levels of HITs but they got stuck at quotient.v and V.v. This PR fixes these issues so it seems that issue is fixed too.

**Comment by mikeshulman** [comment]
Yes, I agree.
